Feature/news sentiment

Sai-132 opened the pull request:
# Description

Newssentiment is a function that displays top news articles and associated sentiment for stock tickers provided by media aggregator Onclusive, through the alternative data platform Althub

- [ ] newssentiment is in the main menu of your terminal

- [ ] newssentiment/show is used to get the news articles and the associated sentiment

- [ ] usage:

        [show, -t,--ticker, -sd, --start_date, -ed, --end_date, -d, --date, -l,--limit,-o,--offset, --export]
        
        These all are optional arguments:

        -t, --ticker [ TICKER ]
                 Ticker to analyze (default: None)
        -sd, --start_date [ START DATE ]
                 The starting date (format YYYY-MM-DD) to search articles from (default: False)
        -ed, --end_date [ END DATE ]
                 The end date (format YYYY-MM-DD) to search articles up to (default: False)
        -d, --date [ DATE ]
                 Show the article data on this day (format YYYY-MM-DD). (default: False)
                 If you use the date argument start_date and end_date arguments will be ignored.
        -l, --limit [ LIMIT ]
                 Number of articles to be displayed (default: 100)
        -o, --offset [ OFFSET ]
                 The offset indicates the starting position of article records (default: 0)
        --export [ EXPORT ]
                 Export raw data into csv, json, xlsx (default: )

        Note: if a date or a date range is not specified, articles from the last five business days are considered.


- [ ] example:
  
        show -t AAPL -sd 2023-01-01 -ed 2023-01-30 -l 50 -o 200

# How has this been tested?

* Please describe the tests that you ran to verify your changes.
* Provide instructions so we can reproduce.
* Please also list any relevant details for your test configuration.
- [ ] Make sure affected commands still run in terminal
- [ ] Ensure the SDK still works
- [ ] Check any related reports


# Checklist:

- [ ] I have adhered to the GitFlow naming convention and my branch name is in the format of `feature/feature-name` or `hotfix/hotfix-name`.
- [ ] Update [our documentation](https://openbb-finance.github.io/OpenBBTerminal/) following [these guidelines](https://github.com/OpenBB-finance/OpenBBTerminal/tree/main/website).  Update any user guides that are affected by the changes.
- [ ] Update our tests following [these guidelines](https://github.com/OpenBB-finance/OpenBBTerminal/tree/main/tests).
- [ ] Make sure you are following our [CONTRIBUTING guidelines](https://github.com/OpenBB-finance/OpenBBTerminal/blob/main/CONTRIBUTING.md).
- [ ] If a feature was added make sure to add it to the corresponding [integration test script](https://github.com/OpenBB-finance/OpenBBTerminal/tree/develop/openbb_terminal/miscellaneous/integration_tests_scripts).


# Others
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.

deeleeramone commented:
Hey, thanks for the PR! A few notes on this:

1. This belongs in the Stocks menu because it is specific to Stocks.
2. This doesn't warrant its own sub-menu, it is one function, and is more suited to be in the `Behavioural Analysis` or `Discovery` sub-menus.
3. Arguments for Terminal functions (argparse) need to follow the convention: `-` for one letter only, `--` for everything else. `-s & --start_date`, `-e & --end_date`.
4. The nested completer is not working for the function's arguments.
5. The name of the function does not correlate with what the function does. `show` is used as an argument elsewhere, a better handle for the function would be: `ns`, short for news_sentiment.
6. Naming needs to be followed. `news_sentiment` instead of `NewsSentiment`. `onclusivedata_controller`, etc. No capitals in the file names; `OnclusiveData` would be used as a class object name, and not a file or function.
7. The folder has no `__init__.py` file, nothing can be imported from the folder. This should be moved to a sub-menu anyways, the model and view files are the only ones which need to travel.
8. Most importantly, It does not appear to work.
![Screenshot 2023-02-27 at 5 10 21 PM](https://user-images.githubusercontent.com/85772166/221726404-8583ea0a-43d6-41db-b1f2-84e3b5d0c080.png)

Sai-132 commented:
As per the instructions above, I have updated the folder structure and created ns (News Sentiment) files under the Discovery sub-menus.

![image](https://user-images.githubusercontent.com/126168589/221891449-1aa06dd5-f4f7-4f4f-9bf3-6dfb1ff02b2a.png)

- To access Althub APIs, you just need to set the environment variable (OPENBB_ALTHUB_API_TOKEN).

- In our previous email conversations, we sent an API_TOKEN. You can access our APIs using that token.

deeleeramone commented:
This has not been added to the `/keys` menu, and instructions in the documentation pages need to be included for obtaining any API token. 

```
To access Althub APIs, you just need to set the environment variable (OPENBB_ALTHUB_API_TOKEN).
```
![Screenshot 2023-02-28 at 8 37 00 AM](https://user-images.githubusercontent.com/85772166/221918171-1191c1ac-76b3-4c57-b4ea-3ba6b60ae9ba.png)

Error when handling no key entered:

```
2023 Feb 28, 11:38 (🦋) /stocks/disc/ $ ns

OPENBB_ALTHUB_API_TOKEN not defined. Set API Keys in ~/.openbb_terminal/.env or under keys menu.
Error: 'NoneType' object has no attribute 'dropna'
```

Please run the testing suite locally and clear all linting errors:

```
************* Module openbb_terminal.stocks.discovery.news_sentiment_model
news_sentiment_model.py 63: C0301 Line too long (144/140) (line-too-long)
news_sentiment_model.py 63: W1309 Using an f-string that does not have any interpolated variables (f-string-without-interpolation)
news_sentiment_model.py 9: C0411 standard import "import datetime" should be placed before "import requests" (wrong-import-order)
news_sentiment_model.py 10: C0411 standard import "import os" should be placed before "import requests" (wrong-import-order)
news_sentiment_model.py 5: W0611 Unused import json (unused-import)
news_sentiment_model.py 8: W0611 Unused numpy imported as np (unused-import)
news_sentiment_model.py 9: W0611 Unused import datetime (unused-import)
news_sentiment_model.py 10: W0611 Unused import os (unused-import)

-----------------------------------
Your code has been rated at 7.89/10
```

deeleeramone commented:
You should also update this branch to be current with `origin/develop`

deeleeramone commented:
Type error in the controller, string is not a bool. 

Please review the help messages for each argument, and correct the grammar.

![Screenshot 2023-02-28 at 8 49 29 AM](https://user-images.githubusercontent.com/85772166/221921383-b3d13caa-74d7-4977-9afd-78e20310e5c5.png)

Sai-132 commented:
I have gone through all the above notes and updated the code accordingly.

I updated the `origin/develop` branch also.

![Screenshot 2023-03-01 171229](https://user-images.githubusercontent.com/126168589/222133962-b7af4cf4-cfb0-403d-9d44-774e95e30d9f.png)

I updated the type for all arguments.

Sai-132 commented:
I have updated the API key also

![image](https://user-images.githubusercontent.com/126168589/222136814-a511d44b-da64-46e7-8e41-7e129b1b7d86.png)

deeleeramone commented:
> I have updated the API key also
> 
> ![image](https://user-images.githubusercontent.com/126168589/222136814-a511d44b-da64-46e7-8e41-7e129b1b7d86.png)

Please resolve the merge conflicts, the branch has not been updated to be in sync with the develop branch until this has happened.

Sai-132 commented:
Can you please check now
I have resolved all the conflicts

deeleeramone commented:
> Can you please check now
> I have resolved all the conflicts

Please resolve all conflicts.

![Screenshot 2023-03-08 at 5 15 23 PM](https://user-images.githubusercontent.com/85772166/223889817-8367fa7e-5465-4da1-b5a0-79b3d857b8d4.png)

Sai-132 commented:
resolved conflicts in the config_terminal file

deeleeramone commented:
> resolved conflicts in the config_terminal file 

Not sure what you did there, but now the keys menu does not work. This error does not occur on the `origin/develop` branch.  There should be no reason for you to have edited this file, and you should revert any changes made. 

![Screenshot 2023-03-09 at 11 16 15 AM](https://user-images.githubusercontent.com/85772166/224131081-93057eef-7d49-494e-b1e4-a3794936fb39.png)

deeleeramone commented on openbb_terminal/sdk.py:
```diff
@@ -20,6 +20,8 @@
     controllers as ctrl,
     models as model,
 )
+from openbb_terminal import feature_flags as obbff
```
Why are these items being added? Line 2 of the file says explicitly that this file is auto generated.

Sai-132 replied:
While Running this Command "python generate_sdk.py sort"
These lines are automatically generated.

deeleeramone commented on openbb_terminal/terminal_controller.py:
```diff
@@ -413,6 +413,7 @@ def call_stocks(self, _):
 
         self.queue = self.load_class(StocksController, self.queue)
 
+
```
Why are we adding an empty new line here?

Sai-132 replied:
This empty new line may have been added by mistake.

Sai-132 commented:
- While Running this Command "python generate_sdk.py sort". These lines are automatically generated.
- The empty line was removed from the terminal_controller file

jmaslek commented on openbb_terminal/config_terminal.py:
```diff
@@ -1,9 +1,6 @@
 import os
 
-# IMPORTATION THIRDPARTY
-import i18n
```
This is not allowing me to test out the terminal:

```
(obb) james@Jamess-MBP OpenBBTerminal % python terminal.py  
Traceback (most recent call last):
  File "/Users/james/OpenBBTerminal/OpenBBTerminal/terminal.py", line 5, in <module>
    import openbb_terminal.config_terminal as cfg  # noqa: F401
  File "/Users/james/OpenBBTerminal/OpenBBTerminal/openbb_terminal/__init__.py", line 13, in <module>
    from .core.plots.plotly_helper import OpenBBFigure, plots_backend, theme, TerminalStyle
  File "/Users/james/OpenBBTerminal/OpenBBTerminal/openbb_terminal/core/plots/__init__.py", line 3, in <module>
    from .plotly_ta.base import Indicator, PltTA, indicator
  File "/Users/james/OpenBBTerminal/OpenBBTerminal/openbb_terminal/core/plots/plotly_ta/base.py", line 5, in <module>
    from .data_classes import ChartIndicators, TAIndicator
  File "/Users/james/OpenBBTerminal/OpenBBTerminal/openbb_terminal/core/plots/plotly_ta/data_classes.py", line 12, in <module>
    from openbb_terminal.common.technical_analysis import ta_helpers
  File "/Users/james/OpenBBTerminal/OpenBBTerminal/openbb_terminal/common/technical_analysis/ta_helpers.py", line 7, in <module>
    from openbb_terminal.rich_config import console
  File "/Users/james/OpenBBTerminal/OpenBBTerminal/openbb_terminal/rich_config.py", line 15, in <module>
    from openbb_terminal import (
  File "/Users/james/OpenBBTerminal/OpenBBTerminal/openbb_terminal/config_terminal.py", line 91, in <module>
    i18n.load_path.append(i18n_dict_location)
NameError: name 'i18n' is not defined
```

jmaslek commented:
It looks like some of the merge conflicts were not correctly resolved.  In the keys model, the information that was stored in the `cfg` is now stored in `current_user.credentials` (check out the keys model in the develop branch).  Most of these changes have been reverted here

Sai-132 commented:
- As per the above comments, I made changes accordingly and resolved conflicts as well.

- Setting API key is working

![image](https://user-images.githubusercontent.com/126168589/225242845-800cd4d6-cda4-4946-b528-fb22f9fd8c25.png)

- API is also working fine

![image](https://user-images.githubusercontent.com/126168589/225244239-40b8cdc5-aa1b-4ac5-973d-3c08fc1cf768.png)

jmaslek commented on openbb_terminal/stocks/discovery/news_sentiment_model.py:
```diff
@@ -0,0 +1,65 @@
+        headers=headers,
+        params=query_params,
+    ).json()
+    df = pd.DataFrame(data=response["results"])
```
This assumes every response will be valid.  When no key is provided, there should be error handling.  For example, without the key defined:

```
API_ALTHUB_TOKEN not defined. Set API Keys in ~/.openbb_terminal/.env or under keys menu.
Error: 'NoneType' object has no attribute 'dropna'
```

jmaslek commented on openbb_terminal/stocks/discovery/news_sentiment_model.py:
```diff
@@ -0,0 +1,65 @@
+    df = pd.DataFrame(data=None)
+
+    query_params = {
+        "all_feilds": 'Flase',
```
Spelling

jmaslek commented on openbb_terminal/stocks/discovery/news_sentiment_model.py:
```diff
@@ -0,0 +1,65 @@
+
+@log_start_end(log=logger)
+@check_api_key(["API_ALTHUB_TOKEN"])
+def get_data(
```
As per contributing guidelines, please include docstrings with all functions.

jmaslek commented on openbb_terminal/stocks/discovery/news_sentiment_view.py:
```diff
@@ -0,0 +1,82 @@
+            console.print()
+            console.print("Date : ", row["published_on"])
+            console.print("Date : ", row["ticker"])
+            console.print("Aritcle name : ", row["article_title"])
+            console.print("Aritcle URL : ", row["article_url"])
+            console.print("Aritcle Sentiment : ", row["raw_sentiment"])
+            console.print("Aritcle Adjusted Sentiment : ", row["adjusted_sentiment"])
+            console.print()
```
Spelling

jmaslek commented on openbb_terminal/stocks/discovery/news_sentiment_view.py:
```diff
@@ -0,0 +1,82 @@
+    if df.empty:
+        console.print("No data found.")
+    else:
+        for index, row in df.iterrows():
```
for _,row in

Undefined index will probably trip a linter

Sai-132 replied:
I didn't get you exactly. Could you please elaborate on it?

jmaslek reviewed (commented):
The trailmap.csv also looks like it was copy pasted over, as things are appearing twice.

Sai-132 commented:
- I updated all the comments you mentioned above.

jmaslek commented:
Thanks.  I have a question about the data we are seeing.  WHen I look at `ns -t aapl`, a couple results are:

```

Date :  2023-03-13T22:46:07Z
Ticker :  AAPL
Article Name :  Article273104510
Article URL :  https://www.heraldonline.com/news/business/article273104510.html
Article Sentiment :  -50.0
Article Adjusted Sentiment :  -360.0


Date :  2023-03-13T22:46:07Z
Ticker :  AAPL
Article Name :  Article F3F70Def 97Bc 505B Bef5 2E6B4551Ac40
Article URL :  https://www.mdjonline.com/arena/thestreet/buffetts-berkshire-hathaway-financial-stocks-slide-on-silicon-valley-bank-fallout/article_f3f70def-97bc-505b-bef5-2e6b4551ac40.html
Article Sentiment :  0.0
Article Adjusted Sentiment :  0.0


Date :  2023-03-13T22:46:07Z
Ticker :  AAPL
Article Name :  Article 56E42D7C D9Eb 5244 8C17 Ac20Eeb027B2
Article URL :  https://www.tribuneledgernews.com/arena/thestreet/buffetts-berkshire-hathaway-financial-stocks-slide-on-silicon-valley-bank-fallout/article_56e42d7c-d9eb-5244-8c17-ac20eeb027b2.html
Article Sentiment :  0.0
Article Adjusted Sentiment :  0.0
```


Why is there no article name on these and 2 out of 3 results show 0 sentiment/adjusted sentiment?

deeleeramone commented:
Attempting to enter API key fails. Are you planning to make this endpoint open to the public? There is currently no way for anyone to sign up and get their own key. 

![Screenshot 2023-03-21 at 10 41 18 PM](https://user-images.githubusercontent.com/85772166/226812796-a88b7b06-f86e-4ee3-8f12-9a5b655f216c.png)

deeleeramone commented:
> Sentiment is classified into three types
>
>  1. Positive :- for this we used 50
>   2. Negative :- for this we used -50
>   3. Natural :- for this we used 0

This really doesn't make much sense. Why not simply, `positive = positive`, `negative = negative`, `neutral = neutral`. Note that the word is not natural, it's neutral. More importantly, how are these conclusions arrived at?  

"Adjusted Sentiment" = "Sentiment" * "Page Rank"

This equation, using the values you have assigned to "Sentiment", will result in lots entries being multiplied by 0. "Page Rank", therefore, has no impact on the "Adjusted Sentiment Score" when "Sentiment" is "Neutral", but does when it is -50 or + 50. 

What does "Adjusted Sentiment Score" even mean? Like, what's a perfect score? What's the worst score you can get?

Sai-132 commented:
> * I have updated it as per the above comments.
> 
> ![image](https://user-images.githubusercontent.com/126168589/227127163-4b3a6ee7-3d4f-45fe-99b5-5efd521d9442.png)


- We fixed new conflicts and updated new settings in feature/news_sentiment

- Please check and confirm if any modification is required.

Sai-132 commented:
> This really doesn't make much sense. Why not simply, `positive = positive`, `negative = negative`, `neutral = neutral`. Note that the word is not natural, it's neutral. More importantly, how are these conclusions arrived at?
> 
> "Adjusted Sentiment" = "Sentiment" * "Page Rank"
> 
> This equation, using the values you have assigned to "Sentiment", will result in lots entries being multiplied by 0. "Page Rank", therefore, has no impact on the "Adjusted Sentiment Score" when "Sentiment" is "Neutral", but does when it is -50 or + 50.
> 
> What does "Adjusted Sentiment Score" even mean? Like, what's a perfect score? What's the worst score you can get? How is page rank quantified as a measurement? Some things may only return 1 page of results only, does that make it better because it's on page 1?

- We have already updated it as per your suggestions.

Please let us konw if you have any comments on this.

deeleeramone commented:
> Please let us konw if you have any comments on this.

Honest question here, the exact same news article is re-published in multiple places, yet receives different scores. How can one be "super positive" and another be "neutral"?  The second part to this question is, which part of this article is "super positive" for AAPL?  The article is specifically talking about a slowing economy in film and television production. 

![Screenshot 2023-04-14 at 8 09 17 PM](https://user-images.githubusercontent.com/85772166/232180156-559b5814-4e2f-4558-96ef-40cc21a0d86e.png)

deeleeramone reviewed (changes requested):
The function should reference the loaded ticker so that a user can simply enter the command as, `ns`, without engaging the `-t` argument.  

To accomplish this, the other functions within the menu do not have a `--ticker` argument because that is handled by the `load` command.

![Screenshot 2023-04-14 at 8 23 58 PM](https://user-images.githubusercontent.com/85772166/232180602-a0588555-6e75-4ffc-a47e-7bdef6692f67.png)

and then instead of `ticker=ns_parser.ticker`, it is set as `ticker=self.ticker`.  For a reference point you can look at the function directly above `call_ns`, in `ba_controller.py`.

![Screenshot 2023-04-14 at 8 25 42 PM](https://user-images.githubusercontent.com/85772166/232180675-ab123105-667e-4638-914e-bca2be378191.png)

deeleeramone reviewed (changes requested):
There is an error when attempting to discover the accepted look-back period.  What is the window of time available for `--start_date` and `--end_date`?

![Screenshot 2023-04-14 at 8 36 09 PM](https://user-images.githubusercontent.com/85772166/232181087-eb3be034-72e4-4b03-8680-3d01bc6ab2f3.png)

Queries with no results - which are not coding errors - should return an error message that helps the user understand what the boundaries, or accepted values, of the parameters are.

deeleeramone reviewed (changes requested):
This documentation for obtaining and authorizing API keys must be updated:

https://docs.openbb.co/terminal/usage/guides/api-keys
https://docs.openbb.co/sdk/usage/guides/api-keys

This function cannot be merged unless there is a method for users to obtain and authorize their own API key, or the data endpoints are opened to the public with no access credentials.

Sai-132 commented:
> > Please let us konw if you have any comments on this.
> 
> Honest question here, the exact same news article is re-published in multiple places, yet receives different scores. How can one be "super positive" and another be "neutral"? The second part to this question is, which part of this article is "super positive" for AAPL? The article is specifically talking about a slowing economy in film and television production.
> 
> ![Screenshot 2023-04-14 at 8 09 17 PM](https://user-images.githubusercontent.com/85772166/232180156-559b5814-4e2f-4558-96ef-40cc21a0d86e.png)

Hi, while the URL title seems very similar, the content might be different as they are published in 2 different sources. Anyway, it's not Althub that calculates the actual sentiment. It is Onclusive who is the owner of the sentiment data. Althub is simply integrating that data with OpenBB. We do a bit of enrichment on the data by calculating adjusted sentiment which is sentiment adjusted as per the trustworthiness of the domain where the article is published.

deeleeramone commented:
> Hi, while the URL title seems very similar, the content might be different as they are published in 2 different sources. 

I know for a fact that the content is identical and the same author is credited.

> We do a bit of enrichment on the data by calculating adjusted sentiment which is sentiment adjusted as per the trustworthiness of the domain where the article is published.

I still am left wondering what that even means and how the exact same article receives different scores.

minhhoang1023 commented:
@Sai-132 Can you help fix the merge conflict so we can get this merge soon? ✨

Sai-132 commented:
Fixed all merge conflicts and it is accessible without an API key.

Sai-132 commented:
@minhhoang1023 We've already updated the code, fixed the conflicts, and your terminal users can access it without an API key.
